src/main/java/org/candlepin/async/tasks/CertificateCleanupJob.java

```diff
@@ -0,0 +1,96 @@
+ * in this software or its documentation.
+ */
+package org.candlepin.async.tasks;
+
+import org.candlepin.async.AsyncJob;
+import org.candlepin.async.JobExecutionContext;
+import org.candlepin.async.JobExecutionException;
+import org.candlepin.model.CertificateSerialCurator;
+import org.candlepin.model.Consumer;
+import org.candlepin.model.ConsumerCurator;
+import org.candlepin.model.ContentAccessCertificate;
+import org.candlepin.model.ContentAccessCertificateCurator;
+import org.candlepin.model.IdentityCertificate;
+import org.candlepin.model.IdentityCertificateCurator;
+
+import com.google.inject.Inject;
+import com.google.inject.persist.Transactional;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Objects;
+
+
+/**
+ * The job will periodically delete all expired identity and content certificates
+ * as well as all revoked and expired certificate serials.
+ */
+public class CertificateCleanupJob implements AsyncJob {
+    private static final Logger log = LoggerFactory.getLogger(CertificateCleanupJob.class);
+
+    public static final String JOB_KEY = "CertificateCleanupJob";
+    public static final String JOB_NAME = "Certificate Cleanup";
+    public static final String DEFAULT_SCHEDULE = "0 0 12 * * ?";
+
+    private final ConsumerCurator consumerCurator;
+    private final IdentityCertificateCurator identCertCurator;
+    private final ContentAccessCertificateCurator caCertCurator;
+    private final CertificateSerialCurator serialCurator;
+
+    @Inject
+    public CertificateCleanupJob(
+        ConsumerCurator consumers,
+        IdentityCertificateCurator identCerts,
+        ContentAccessCertificateCurator contentAccessCerts,
+        CertificateSerialCurator serialCurator) {
+        this.consumerCurator = Objects.requireNonNull(consumers);
+        this.identCertCurator = Objects.requireNonNull(identCerts);
+        this.caCertCurator = Objects.requireNonNull(contentAccessCerts);
+        this.serialCurator = Objects.requireNonNull(serialCurator);
+    }
+
+    @Override
+    @Transactional
+    public void execute(JobExecutionContext context) throws JobExecutionException {
+        log.debug("Starting certificate cleanup.");
+        // Skipping entitlement certificates as those are handled by ExpiredPoolsCleanupJob
+        cleanupExpiredIdentityCerts();
+        cleanupExpiredContentAccessCerts();
+        this.serialCurator.deleteRevokedExpiredSerials();
+        log.debug("Certificate cleanup successfully completed.");
+    }
+
+    private void cleanupExpiredIdentityCerts() {
+        List<Consumer> consumers = this.consumerCurator.findAllWithExpiredIdCerts();
+        log.info("Cleaning up {} expired identity certificates.", consumers.size());
+        for (Consumer consumer : consumers) {
+            IdentityCertificate idCert = consumer.getIdCert();
+            consumer.setIdCert(null);
+            this.identCertCurator.delete(idCert);
```

Comment:
There would be problems here the first time this job would run in Hosted, due to millions of expired id certs lying around, so we should re-think this approach.

Instead of fetching potentially millions of Consumers into memory for the sole purpose of deleting their id certs, we should be doing all these operations through a few queries instead, e.g.:
1. Fetch the serials of id certs that are expired
2. Update cp_consumer.consumer_idcert_id to NULL for those certs who have those serials
3. Delete the cp_id_cert rows for those serials
4. Update the cp_cert_serial to revoked=true for those serials
Also, since I suspect these queries would make use of the IN clause, we should be partitioning the cert serials into batches like we do in other Curators, to guard against dbms limitations there.

Reply:
Good point. I did not considered a scenario where is accumulated so much already. Done